### PR TITLE
add helm tests

### PR DESCRIPTION
**Pull request by nieuwsma**
## Summary and Scope

Updates the helm chart to have a built in helm test. This replaces the CT tests in the RPMs.

## Testing

_List the environments in which these changes were tested._

Tested on:

  * mug

Test description:

Test process.  I dont care about FAS data on mug so I nuked it.

```
helm -n services uninstall cray-hms-firmware-action
wget https://artifactory.algol60.net/artifactory/csm-helm-charts/unstable/cray-hms-firmware-action/cray-hms-firmware-action-2.1.0-20220301183128+bc8891e.tgz
# HAVE TO MAKE SURE THE IMAGES ARE ON THE SYSTEM!!!!
helm upgrade --install -n services cray-hms-firmware-action ./cray-hms-firmware-action-2.1.0-20220301183128+bc8891e.tgz
helm test -n services cray-hms-firmware-action 
```

```
ncn-m001:~/anieuwsma # helm test -n services cray-hms-firmware-action
NAME: cray-hms-firmware-action
LAST DEPLOYED: Tue Mar  1 18:22:35 2022
NAMESPACE: services
STATUS: deployed
REVISION: 1
TEST SUITE:     cray-hms-firmware-action-test-functional
Last Started:   Tue Mar  1 18:28:08 2022
Last Completed: Tue Mar  1 18:30:50 2022
Phase:          Succeeded
TEST SUITE:     cray-hms-firmware-action-test-smoke
Last Started:   Tue Mar  1 18:30:50 2022
Last Completed: Tue Mar  1 18:32:08 2022
Phase:          Succeeded
```

```
ncn-m001:~/anieuwsma # kubectl -n services logs cray-hms-firmware-action-test-smoke-4bfc2 smoke
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to localhost port 15000 after 1 ms: Connection refused
Waiting for Sidecar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
HTTP/1.1 200 OK
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
content-type: text/html; charset=UTF-8
cache-control: no-cache, max-age=0
x-content-type-options: nosniff
date: Tue, 01 Mar 2022 17:03:52 GMT
server: envoy
transfer-encoding: chunked

Sidecar available
Running smoke tests...
2022-03-01 17:03:52,658 suite_name: FAS-smoke-tests
2022-03-01 17:03:52,658 file_path: smoke.json
2022-03-01 17:03:52,658 Running FAS-smoke-tests...
2022-03-01 17:03:52,658 default_url: http://cray-fas
2022-03-01 17:03:52,658 override_url: http://cray-fas
2022-03-01 17:03:52,658 configured_url: http://cray-fas/
2022-03-01 17:03:52,658 Testing {"path": "service/version", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/service/version"}
2022-03-01 17:03:52,679 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,743 http://cray-fas:80 "GET /service/version HTTP/1.1" 200 27
2022-03-01 17:03:52,745 Testing {"path": "service/status", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/service/status"}
2022-03-01 17:03:52,782 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,785 http://cray-fas:80 "GET /service/status HTTP/1.1" 200 27
2022-03-01 17:03:52,786 Testing {"path": "service/status/details", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/service/status/details"}
2022-03-01 17:03:52,805 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,819 http://cray-fas:80 "GET /service/status/details HTTP/1.1" 200 105
2022-03-01 17:03:52,820 Testing {"path": "actions", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/actions"}
2022-03-01 17:03:52,844 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,848 http://cray-fas:80 "GET /actions HTTP/1.1" 200 14
2022-03-01 17:03:52,848 Testing {"path": "images", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/images"}
2022-03-01 17:03:52,870 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,881 http://cray-fas:80 "GET /images HTTP/1.1" 200 13
2022-03-01 17:03:52,882 Testing {"path": "snapshots", "expected_status_code": 200, "method": "GET", "body": null, "headers": {}, "url": "http://cray-fas/snapshots"}
2022-03-01 17:03:52,903 Starting new HTTP connection (1): cray-fas:80
2022-03-01 17:03:52,907 http://cray-fas:80 "GET /snapshots HTTP/1.1" 200 16
2022-03-01 17:03:52,907 MAIN_ERRORS = 0
2022-03-01 17:03:52,907 PASS: FAS-smoke-tests
2022-03-01 17:03:52,907 passed!
2022-03-01 17:03:52,907 PASS: FAS-smoke-tests passed!
```

```
ncn-m001:~/anieuwsma # kubectl -n services logs cray-hms-firmware-action-test-functional-xwlw4 functional
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (7) Failed to connect to localhost port 15000 after 1 ms: Connection refused
Waiting for Sidecar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
HTTP/1.1 200 OK
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
content-type: text/html; charset=UTF-8
cache-control: no-cache, max-age=0
x-content-type-options: nosniff
date: Tue, 01 Mar 2022 18:28:15 GMT
server: envoy
transfer-encoding: chunked

Sidecar available
Running functional tests...
============================= test session starts ==============================
platform linux -- Python 3.9.7, pytest-6.1.2, py-1.11.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /src/app, configfile: pytest.ini
plugins: tap-3.3, tavern-1.12.2
collecting ... collected 6 items

test_actions.tavern.yaml::Verify the actions resource PASSED             [ 16%]
test_service_status.tavern.yaml::Verify the service status resource PASSED [ 33%]
test_service_status_details.tavern.yaml::Verify the service status details resource PASSED [ 50%]
test_service_version.tavern.yaml::Verify the service version resource PASSED [ 66%]
test_snapshots.tavern.yaml::Verify the snapshots resource for creating a new snapshot PASSED [ 83%]
test_snapshots.tavern.yaml::Verify the snapshots resource for deleting a snapshot PASSED [100%]

======================== 6 passed in 136.22s (0:02:16) =========================
```

- Were continuous integration tests run? y
- Was upgrade tested? y
- Was downgrade tested? y

### CODE REVIEW

File: ct.yaml [hunk 1]
```diff
@@ -1,7 +1,10 @@
----
-chart-dirs: []
+chart-dirs:
```
Comment (rsjostrand-hpe) [inline ct.yaml:1]: Revert your changes to this file.

**Comment by nieuwsma** [comment]
Relevant process: https://connect.us.cray.com/confluence/display/CSMRBOOKS/Testing+a+Helm+chart+on+a+system+1.4+and+1.5

File: charts/v2.1/cray-hms-firmware-action/values.yaml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+image:
+  repository: artifactory.algol60.net/csm-docker/stable/cray-firmware-action
+  pullPolicy: IfNotPresent
+nexus:
+  endpoint: http://nexus.nexus.svc.cluster.local
+  repo: shasta-firmware
+tests:
+  image:
+    repository: artifactory.algol60.net/csm-docker/stable/cray-firmware-action-test
+    pullPolicy: Always
```
Comment (schooler-hpe) [inline charts/v2.1/cray-hms-firmware-action/values.yaml:21]: Did you want to update this to 'IfNotPresent'?

**Comment by schooler-hpe** [comment]
My only other comment would be to change the order in which the tests run so that the smoke tests run first, then the functional. Even better if we don't execute the functional tests if there is a smoke test failure first but it sounds like that requires 'helm test --filter' in Helm v3.6.0 which we don't yet have.

**Comment by nieuwsma** [comment]
> My only other comment would be to change the order in which the tests run so that the smoke tests run first, then the functional. Even better if we don't execute the functional tests if there is a smoke test failure first but it sounds like that requires 'helm test --filter' in Helm v3.6.0 which we don't yet have.

well there are a few things we could maybe do...
I could change the job name so smoke is `asmoke`  so that it comes before hand... but im not sure that would stop the second job.

Maybe I could make it one big job... but then id have to rework the entrypoint script.

**Comment by rsjostrand-hpe** [comment]
@nieuwsma to get the smoke tests to run first I would maybe try to look at the `helm.sh/hook-weight` annotation, this should allow you to weight the different Helm hooks to run a in particular order.  SLS has an example of using the hook-weight annotation: https://github.com/Cray-HPE/hms-sls-charts/blob/89e9b61f87eb3f0131d5dda3f6b9b8374e7c0ed9/charts/v2.0/cray-hms-sls/templates/jobs.yaml

This page https://helm.sh/docs/topics/chart_tests/ mentions that the `helm.sh/hook-weight` annotation also works for tests.

**Comment by nieuwsma** [comment]
> @nieuwsma to get the smoke tests to run first I would maybe try to look at the `helm.sh/hook-weight` annotation, this should allow you to weight the different Helm hooks to run a in particular order. SLS has an example of using the hook-weight annotation: https://github.com/Cray-HPE/hms-sls-charts/blob/89e9b61f87eb3f0131d5dda3f6b9b8374e7c0ed9/charts/v2.0/cray-hms-sls/templates/jobs.yaml
> 
> This page https://helm.sh/docs/topics/chart_tests/ mentions that the `helm.sh/hook-weight` annotation also works for tests. Also this section has some info about it: https://helm.sh/docs/topics/charts_hooks/#writing-a-hook

Fixed

**Review by rsjostrand-hpe — APPROVED** [review]

**Comment by nieuwsma** [comment]
Got the helm weights working:

```NAME: cray-hms-firmware-action
LAST DEPLOYED: Wed Mar  2 16:18:17 2022
NAMESPACE: services
STATUS: deployed
REVISION: 1
TEST SUITE:     cray-hms-firmware-action-test-functional
Last Started:   Wed Mar  2 16:26:29 2022
Last Completed: Wed Mar  2 16:30:05 2022
Phase:          Succeeded
TEST SUITE:     cray-hms-firmware-action-test-smoke
Last Started:   Wed Mar  2 16:25:30 2022
Last Completed: Wed Mar  2 16:26:29 2022
Phase:          Succeeded```

**Pull request merged**